src/coprocessor/util.rs

```diff
@@ -17,33 +17,89 @@ use tipb::schema::ColumnInfo;
 use super::codec::datum::Datum;
 use super::codec::mysql;
 
-/// Get the smallest key which is larger than the key given.
-pub fn prefix_next(key: &[u8]) -> Vec<u8> {
-    let mut nk = key.to_vec();
-    if nk.is_empty() {
-        nk.push(0);
-        return nk;
+/// Convert the key to the smallest key which is larger than the key given.
+pub fn convert_to_prefix_next(key: &mut Vec<u8>) {
+    if key.is_empty() {
+        key.push(0);
+        return;
     }
-    let mut i = nk.len() - 1;
+    let mut i = key.len() - 1;
+
+    // Add 1 to the last byte that is not 255, and set it's following bytes to 0.
     loop {
-        if nk[i] == 255 {
-            nk[i] = 0;
+        if key[i] == 255 {
+            key[i] = 0;
         } else {
-            nk[i] += 1;
-            return nk;
+            key[i] += 1;
+            return;
         }
         if i == 0 {
-            nk = key.to_vec();
-            nk.push(0);
-            return nk;
+            // All bytes are 255. Append a 0 to the key.
+            for byte in key.iter_mut() {
+                *byte = 255;
+            }
+            key.push(0);
+            return;
         }
         i -= 1;
     }
 }
 
+/// Check if `key`'s prefix next equals to `next`
+pub fn is_prefix_next(key: &[u8], next: &[u8]) -> bool {
+    let len = key.len();
+    let next_len = next.len();
+
+    if len == next_len {
+        // Find the last non-255 byte
+        let mut carry_pos = len;
```

Comment:
This loop might be written in a more understandable way?

```
for carry_pos in (0..len-1).rev() {
    if key[carry_pos] != 255 {
        break;
    }
}
if carry_pos == 0 {
    // All bytes of `key` are 255. `next` couldn't be `key`'s prefix_next since their
    // lengths are equal.
    return false;
}
```

Reply:
`carry_pos` in the for loop hides the outer one.
Do you think this code is better?
```
        // Find the last non-255 byte
        // usize::max_value() means no byte is 255.
        let mut carry_pos = usize::max_value();
        for i in (0..len).rev() {
            if key[i] != 255 {
                carry_pos = i;
                break;
            }
        }
        
        if carry_pos == usize::max_value() {
            // All bytes of `key` are 255. `next` couldn't be `key`'s prefix_next since their
            // lengths are equal.
            return false;
        }
```

Reply:
oh. `usize::max_value()`.. seems tricky. it's the value of -1 then..

Reply:
let's keep it here.